### ISSUE TITLE
Suggestion: script to fix database after DB server crash or unplanned reboot?

### ISSUE THREAD

**Issue by ghost** [cryptocoinexplorer/Open-Source-CCE-4.0#4]
Hello

I've a small suggestion following an unplanned DB server reboot by my VPS provider.
Apparently this happened while CCE was in the middle of writing a bunch a records to the database.
Since I wasn't notified, I only saw that no more data was being added in the past 11 days.

Checking the logs showed a load of lines such as these:

```
ERROR:root:04-13 17:24:21 (1062, u"Duplicate entry '4b28de900892da7b404a0146f0ca
6326b29ce1f4123a35eb22e210bec8ef347b' for key 'tx_hash'") : query no return
ERROR:root:04-13 17:24:21 INSERT INTO tx_raw (tx_hash,raw,decoded,height) VALUES
(%s,%s,%s,%s) : query no return
ERROR:root:04-13 17:24:21 (1062, u"Duplicate entry 'acc4c48a0de375f54d281483d75e
8f91b5fff327ad31b932d5b9e19d244870ce' for key 'tx_hash'") : query no return
ERROR:root:04-13 17:24:21 INSERT INTO tx_raw (tx_hash,raw,decoded,height) VALUES
(%s,%s,%s,%s) : query no return
ERROR:root:04-13 17:24:21 (1062, u"Duplicate entry '3041839' for key 'PRIMARY'")
 : Main loop
ERROR:root:04-13 17:25:58 (1062, u"Duplicate entry '98de8a677c4f259fa6697b1a4c0d
24712773c10e79cbaa92ed56830709710caf' for key 'tx_hash'") : query no return
ERROR:root:04-13 17:25:58 INSERT INTO tx_raw (tx_hash,raw,decoded,height) VALUES
(%s,%s,%s,%s) : query no return
ERROR:root:04-13 17:25:58 (1062, u"Duplicate entry 'ca541cc26aa3d7eeb23b51e85ca1
224fcd18c2f72a7c4e1497bbc0fcd5381a92' for key 'tx_hash'") : query no return
ERROR:root:04-13 17:25:58 INSERT INTO tx_raw (tx_hash,raw,decoded,height) VALUES
(%s,%s,%s,%s) : query no return
ERROR:root:04-13 17:25:58 (1062, u"Duplicate entry '3041873' for key 'PRIMARY'")
 : Main loop
ERROR:root:04-13 17:26:36 DBLoader Timeout : jsonrpc
ERROR:root:04-13 17:26:36 sequence index must be integer, not 'str' : Main loop
ERROR:root:04-13 17:27:27 (1062, u"Duplicate entry '294eb67cd660497910beb7fd28a1
b1cc8793fc0ce9d24493061a3f54ca2c0d0f' for key 'tx_hash'") : query no return
ERROR:root:04-13 17:27:27 INSERT INTO tx_raw (tx_hash,raw,decoded,height) VALUES
(%s,%s,%s,%s) : query no return
ERROR:root:04-13 17:27:27 (1062, u"Duplicate entry '9b11bd5c2ccddff329c8f046ce28
2b5c9f4a5c8d594c779144e36fe580c3c9b7' for key 'tx_hash'") : query no return
ERROR:root:04-13 17:27:27 INSERT INTO tx_raw (tx_hash,raw,decoded,height) VALUES
(%s,%s,%s,%s) : query no return
ERROR:root:04-13 17:27:27 (1062, u"Duplicate entry '3041910' for key 'PRIMARY'")
 : Main loop
ERROR:root:04-13 17:29:10 (1062, u"Duplicate entry 'ca15ea6de60a30662b2b660a692f
7f78c95e6ea638dc818b809d47ffc63b82ed' for key 'tx_hash'") : query no return
ERROR:root:04-13 17:29:10 INSERT INTO tx_raw (tx_hash,raw,decoded,height) VALUES
(%s,%s,%s,%s) : query no return
ERROR:root:04-13 17:29:10 (1062, u"Duplicate entry '5869edd81ec6a625157e33655193
d94daa7a11556a0f512894f774c221f38c4d' for key 'tx_hash'") : query no return
ERROR:root:04-13 17:29:10 INSERT INTO tx_raw (tx_hash,raw,decoded,height) VALUES
(%s,%s,%s,%s) : query no return
ERROR:root:04-13 17:29:10 (1062, u"Duplicate entry '3041922' for key 'PRIMARY'")
 : Main loop
ERROR:root:04-13 17:31:27 (1062, u"Duplicate entry '9d60012340e835707d6be8c1141b
593cfbf470ee58f22c4b46dfe01e5cef1771' for key 'tx_hash'") : query no return
ERROR:root:04-13 17:31:27 INSERT INTO tx_raw (tx_hash,raw,decoded,height) VALUES
(%s,%s,%s,%s) : query no return
ERROR:root:04-13 17:31:28 (1062, u"Duplicate entry 'f6b14cca3dde6b3117390250b200
7033200c7afffd16fdf2550de569775bc2df' for key 'tx_hash'") : query no return
ERROR:root:04-13 17:31:28 INSERT INTO tx_raw (tx_hash,raw,decoded,height) VALUES
(%s,%s,%s,%s) : query no return
ERROR:root:04-13 17:31:28 (1062, u"Duplicate entry '3041942' for key 'PRIMARY'")
```

Rebooting the VPS and restarting the different associated crypto wallet services didn't help.

I ended up checking for these transactions in the database and identified the first tx_hash that started the trouble. Taking a rather big gamble, I decided to remove everything from the tables that referred to this tx_hash and all those that followed.

Basically, I executed the following SQL commands:

```
delete a
from tx_out a
join tx_raw as b on b.tx_hash = a.tx_hash
where b.height >= 3041839;

delete a 
from tx_in a 
join tx_raw as b on b.tx_hash = a.tx_hash 
where b.height >= 3041839;

delete from tx_raw 
where height >= 3041839;

delete from block 
where block.height >= 3041839;
```

No bad records were found in the orph_ tables.

I might have missed a table with residual data, but so far everything is fine and CCE started reloading the block chain from block 3041839 without any error.

Would it be interesting to write a emergency repair script based on the "solution" described above?
I don't think many users are thrilled to reload 28GB or more of data when their server crash.

Best regards,
Akira

**Comment by ghost** [comment]
Damn! That didn't solve the issue. After a while it's back.
This time, I disabled coin.sh being automatically loaded and ran dbload.py -v manually.
It gets consecutively interrupted with two kinds of error:

```
akira@Mineralscoin:~/CCE$ ./dbload.py -v
Recheck Called
Processing Block:  3041921  of  3082686 
Main Loop sequence index must be integer, not 'str'
akira@Mineralscoin:~/CCE$ ./dbload.py -v
Recheck Called
Processing Block:  3041993  of  3082700 
Main Loop cannot concatenate 'str' and 'int' objects
```

These errors repeat every time in this that order no matter how often I relaunch the process.
